### PR TITLE
APT-10304 Range Setting Issue Resolved

### PR DESCRIPTION
**Pull request by iurii-provizio**
> [!NOTE]
> Introduce stable match/timeout-based request/response readiness (C++/Python), expose match-count helpers, improve error propagation, add reliability tests, handle empty PointCloud2, update CI runners, and make versioning deterministic.
> 
> - **Request/Response (C++/Python)**:
>   - Add stable matching window and optional matching timeout to `request(...)` (replacing post-match delay).
>   - Client waits for equal publisher/subscriber match counts; deferred requests auto-flush on readiness.
>   - Propagate errors via exceptions (`timeout`, `interrupted`, `publish-failed`) and through `future_response::wait_until`.
> - **Publisher/Subscriber**:
>   - Add `get_num_matched_subscribers(...)` and `get_num_matched_publishers(...)` with condition-based tracking; refactor listeners to maintain counts.
> - **Python bindings**:
>   - Mirror new matching/timeout APIs and errors; expose match-count getters; internal readiness logic updated.
> - **PointCloud2**:
>   - Return empty NumPy array when cloud data is empty; minor dtype reuse.
> - **Tests**:
>   - Add reliability tests (C++/Python) and `test/run_times.sh`; adjust dual-service and concurrent tests/timeouts.
> - **Build/CI**:
>   - CI: switch Jetson runner to `jetson-20.04` in test matrices.
>   - CMake: derive version deterministically from HEAD commit timestamp.
>   - install script: make CMake install step non-fatal if already installed.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 0c28b540eea3abe8a424ec7ece06138ec611632c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: include/provizio/dds/publisher.h [hunk 1]
```diff
@@ -402,39 +416,38 @@ namespace provizio::dds
...
+                    // No change during the settle_time period
+                    return num_matched_subscribers;
                 }
-            }
-            else
-            {
-                has_stable_since = false;
-            }
-            std::this_thread::sleep_for(iteration_sleep);
+            } while (std::chrono::steady_clock::now() < timeout_point);
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply include/provizio/dds/publisher.h:444]: Not in this function, timeout of 0 here stands for a single attempt, see the doxygen doc. False positive.

File: python/provizio_dds.py [hunk 2]
```diff
@@ -95,6 +95,54 @@ def __init__(self, pub_sub_type: TopicDataType):
...
+                    return -1
+                wait_timeout = min(settle_time_sec, remaining)
+            else:
+                wait_timeout = settle_time_sec
+
+            def _count_changed(prev=previous):
+                return accessor() != prev
+
+            if not condition.wait_for(_count_changed, timeout=wait_timeout):
+                return previous
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply python/provizio_dds.py:128]: Resolved

File: include/provizio/dds/request_response_details.h [hunk 3]
```diff
@@ -478,50 +577,66 @@ namespace provizio::dds::detail
...
                 {
-                    lock.unlock();
-                    std::this_thread::sleep_for(post_match_delay);
-                    lock.lock();
-                    if (stop)
+                    const auto remaining_timeout = service_match_timeout.count() == 0
+                                                       ? service_match_timeout
+                                                       : std::max(std::chrono::duration_cast<std::chrono::milliseconds>(
+                                                                      std::chrono::steady_clock::now() - timeout_point),
+                                                                  std::chrono::milliseconds{0});
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply include/provizio/dds/request_response_details.h:598]: Fixed

File: include/provizio/dds/request_response_details.h [hunk 4]
```diff
@@ -478,50 +577,61 @@ namespace provizio::dds::detail
...
-                    lock.unlock();
-                    std::this_thread::sleep_for(post_match_delay);
-                    lock.lock();
-                    if (stop)
+                    auto matched_subscribers = std::async(std::launch::async, [&] {
+                        return publisher->get_num_matched_subscribers(service_match_timeout, check_period);
+                    });
+                    auto matched_publishers = std::async(std::launch::async, [&] {
+                        return subscriber->get_num_matched_publishers(service_match_timeout, check_period);
+                    });
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply include/provizio/dds/request_response_details.h:604]: Fixed

File: test/python/python_request_response_reliability_client.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    log_prefix = f"python_request_response_reliability_client{test_name_postfix}: "
+    service_name = (
+        f"provizio_dds_test_request_response_reliability{test_name_postfix}"
+    )
+    domain_id = 14
+    timeout_sec = 5.0
+
+    random.seed(value)
+    wait_ms = random.randint(0, 1999)
```
Comment (ClementBotty) [inline test/python/python_request_response_reliability_client.py]:
```suggestion
    wait_ms = random.randint(0, MAX_WAIT_MS)
```

With MAX_WAIT_MS defined above.

File: test/python/python_request_response_reliability_client.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        print(f"{log_prefix}{_timestamp()}Timeout waiting for response")
+        return 1
+
+    received_value = response.data()
+    if received_value != value:
+        print(
+            f"{log_prefix}{_timestamp()}Unexpected response {received_value}, "
+            f"expected {value}"
+        )
+        return 1
```
Comment (ClementBotty) [inline test/python/python_request_response_reliability_client.py:86]: In general, I would recommend using a python logger to then use logging.info, warning, error, etc. instead of print().

File: test/python/python_request_response_reliability_service.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    log_prefix = f"python_request_response_reliability_service{test_name_postfix}: "
+    service_name = (
+        f"provizio_dds_test_request_response_reliability{test_name_postfix}"
+    )
+    domain_id = 14
+
+    random.seed(expected_value)
+    wait_ms = random.randint(0, 1999)
+    if wait_ms >= 1000:
+        print(f"{log_prefix}{_timestamp()}Waiting {wait_ms}ms...")
```
Comment (ClementBotty) [inline test/python/python_request_response_reliability_service.py:57]: This can be done automatically with a proper logger, instead of having to define the log_prefix.

Comment (ClementBotty) [reply test/python/python_request_response_reliability_service.py:57]: @iurii-provizio See Customer Formater for logging here : https://github.com/provizio/ai_perception_microservices/blob/develop/src/utils/logger.py#L12

File: test/python/python_request_response_reliability_service.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                lambda: requests_processed, timeout=wait_timeout
+            )
+        if not finished:
+            print(f"{log_prefix}{_timestamp()}Timeout waiting for request")
+            return 1
+
+        if not received_expected_values:
+            return 1
+
+        time.sleep(0.25)
```
Comment (ClementBotty) [inline test/python/python_request_response_reliability_service.py]: I'd say it's a magic number here.

File: include/provizio/dds/subscriber.h [hunk 9]
```diff
@@ -80,18 +107,17 @@ namespace provizio::dds
 
         /**
          * @brief Blocks until this subscriber has at least one stable match for a short settle window.
-         * @param timeout Total timeout duration.
+         * @param timeout Total timeout duration, takes a single attempt when 0.
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply include/provizio/dds/subscriber.h:110]: Waiting `min_attempt_time` is what makes it an "attempt", so the description is all correct.

File: include/provizio/dds/publisher.h [hunk 10]
```diff
@@ -186,18 +186,21 @@ namespace provizio::dds
 
         /**
          * @brief Blocks until this publisher has at least one stable match for a short settle window.
-         * @param timeout Total timeout duration.
+         * @param timeout Total timeout duration, takes a single attempt when 0.
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply include/provizio/dds/publisher.h:189]: Waiting min_attempt_time is what makes it an "attempt", so the description is all correct.

File: python/provizio_dds.py [hunk 11]
```diff
@@ -95,6 +95,49 @@ def __init__(self, pub_sub_type: TopicDataType):
...
+
+
+def _get_stable_match_count(
+    condition: threading.Condition,
+    accessor: Callable[[], int],
+    timeout_sec: float,
+    settle_time_sec: float,
+) -> int:
+    settle_time_sec = max(0.0, settle_time_sec)
+    timeout_sec = max(0.0 if timeout_sec is None else timeout_sec, 0.0)
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply python/provizio_dds.py]: Done

File: python/provizio_dds.py [hunk 12]
```diff
@@ -581,34 +639,64 @@ def on_response(data, info):
...
+                None,
+                lambda: request_publisher.get_num_matched_subscribers(remaining, check_period),
+            )
+            sub_task = loop.run_in_executor(
+                None,
+                lambda: response_subscriber.get_num_matched_publishers(remaining, check_period),
+            )
+            num_subscribers, num_publishers = await asyncio.gather(pub_task, sub_task)
+
+            if num_subscribers > 0 and num_subscribers == num_publishers:
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply python/provizio_dds.py]: Done

File: include/provizio/dds/request_response_details.h [hunk 13]
```diff
@@ -530,58 +645,70 @@ namespace provizio::dds::detail
...
+            if (publisher->publish(request_data, params))
+            {
+                context.identity.writer_guid() = subscriber_guid;
+                context.identity.sequence_number() = params.sample_identity().sequence_number();
+                success = true;
+            }
+        }
+
+        if (!success)
         {
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply include/provizio/dds/request_response_details.h:699]: Done

File: test/python/python_request_response_reliability_client.py (hunk 5 above)
Comment (iurii-provizio) [reply test/python/python_request_response_reliability_client.py]: Done

File: test/python/python_request_response_reliability_client.py (hunk 6 above)
Comment (iurii-provizio) [reply test/python/python_request_response_reliability_client.py:86]: @ClementBotty Does it matter in unit tests, which it is here?

File: test/python/python_request_response_reliability_service.py (hunk 7 above)
Comment (iurii-provizio) [reply test/python/python_request_response_reliability_service.py:57]: Sure, but same comment - it's a unit test, my intention was to keep it as simple as it gets and only test the functionality the test intends to

File: python/provizio_dds.py [hunk 14]
```diff
@@ -581,34 +639,67 @@ def on_response(data, info):
...
-    params = WriteParams()
-    params.related_sample_identity().writer_guid(response_subscriber.get_guid())
+            pub_task = loop.run_in_executor(
+                None,
+                lambda: request_publisher.get_num_matched_subscribers(remaining, check_period),
+            )
+            sub_task = loop.run_in_executor(
+                None,
+                lambda: response_subscriber.get_num_matched_publishers(remaining, check_period),
+            )
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply python/provizio_dds.py:671]: Not needed here, as both lambdas only exist in a single iteration and then waited to finish all calls to with `await asyncio.gather` below before proceeding, `remaining` never changes within the lambdas' life time.

File: test/python/python_request_response_reliability_service.py (hunk 8 above)
Comment (iurii-provizio) [reply test/python/python_request_response_reliability_service.py]: Fixed

File: python/point_cloud2.py [hunk 15]
```diff
@@ -109,6 +109,9 @@ def read_points(
     """
     assert isinstance(cloud, PointCloud2), "Cloud is not a provizio_dds.PointCloud2"
 
+    if cloud.data() is None or cloud.data().size() == 0:
+        return np.empty(0)
```
[the remark this answers is not in the document]
Comment (iurii-provizio) [reply python/point_cloud2.py]: Fixed

File: test/python/python_request_response_reliability_client.py (hunk 6 above)
Comment (ClementBotty) [reply test/python/python_request_response_reliability_client.py:86]: @iurii-provizio Not much in unit test indeed. All good then 👍

File: test/python/python_request_response_reliability_service.py (hunk 7 above)
Comment (ClementBotty) [reply test/python/python_request_response_reliability_service.py:57]: All good 👍

**Review by ClementBotty — APPROVED** [review]
All good now !
